### PR TITLE
Add CPU model detection for RISC-V systems

### PR DESCRIPTION
**Pull request by RossComputerGuy**
Partial fix for #18848

Will test later today.

### CODE REVIEW

**Comment by nektro** [comment]
full list appears to be here: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/riscv/cpus.yaml#n30

**Comment by RossComputerGuy** [comment]
@nektro Thank you for the list, I'll add those in and rename this PR.

**Comment by RossComputerGuy** [comment]
```
thread 1787 panic: reached unreachable code
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/std/Target.zig:299:29: 0x115322f in default (build)
                    else => unreachable,
                            ^
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/std/Target.zig:97:54: 0x10ea12d in defaultVersionRange (build)
                .version_range = VersionRange.default(tag, arch),
                                                     ^
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/std/zig/system.zig:167:46: 0x10d2c15 in resolveTargetQuery (build)
    var os = query_os_tag.defaultVersionRange(query.cpu_arch orelse builtin.cpu.arch);
                                             ^
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/std/Build.zig:2393:52: 0x10f2093 in resolveTargetQuery (build)
        .result = std.zig.system.resolveTargetQuery(query) catch
                                                   ^
/home/user/ExpidusOS/zig/test/link/macho.zig:6:48: 0x12cd389 in testAll (build)
    const default_target = b.resolveTargetQuery(.{
                                               ^
/home/user/ExpidusOS/zig/test/link/link.zig:16:52: 0x128f765 in build (build)
    test_step.dependOn(@import("macho.zig").testAll(b, build_opts));
                                                   ^
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/std/Build.zig:1982:33: 0x1240585 in runBuild__anon_24823 (build)
        .Void => build_zig.build(b),
                                ^
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/std/Build.zig:1963:29: 0x11d29c7 in dependencyInner__anon_21926 (build)
        sub_builder.runBuild(bz) catch @panic("unhandled error");
                            ^
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/std/Build.zig:1886:27: 0x1175c79 in anonymousDependency__anon_20078 (build)
    return dependencyInner(b, name, build_root, build_zig, &.{}, args);
                          ^
/home/user/ExpidusOS/zig/test/tests.zig:754:46: 0x110a567 in addLinkTests (build)
            const dep = b.anonymousDependency(case.build_root, case.import, .{
                                             ^
/home/user/ExpidusOS/zig/build.zig:515:42: 0x10fb5cf in build (build)
    test_step.dependOn(tests.addLinkTests(b, enable_macos_sdk, enable_ios_sdk, false, enable_symlinks_windows));
                                         ^
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/std/Build.zig:1983:43: 0x10d4461 in runBuild__anon_8902 (build)
        .ErrorUnion => try build_zig.build(b),
                                          ^
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/build_runner.zig:310:29: 0x10d070f in main (build)
        try builder.runBuild(root);
                            ^
/home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/lib/std/start.zig:585:37: 0x10bc9ab in posixCallMainAndExit (build)
            const result = root.main() catch |err| {
                                    ^
???:?:?: 0x0 in ??? (???)
error: the following build command crashed:
/home/user/ExpidusOS/zig/zig-cache/o/647f9bdf3a96d777087cc5627ffb5555/build /home/user/Downloads/zig-linux-riscv64-0.12.0-dev.2631+3069669bc/zig /home/user/ExpidusOS/zig /home/user/ExpidusOS/zig/zig-cache /home/user/.cache/zig --seed 0x49969672 -Z7c954b624bd4ad1f

real    10m28.858s
user    10m26.625s
sys     0m5.020s
```
Tried building this locally but ran into an issue, guess I'll have to bootstrap on another system and copy it over.

**Comment by RossComputerGuy** [comment]
image
This works.

**Review by Vexu — APPROVED** [review]
Do you want to add those other CPUs before this is merged?

**Comment by RossComputerGuy** [comment]
I already added what I can, the others don't appear to be in `Target.riscv.cpu`.

**Comment by Vexu** [comment]
You can add `extra_cpus` in `tools/update_cpu_features.zig` and then run it agains LLVM 17 source.

**Comment by RossComputerGuy** [comment]
LLVM CPU names vs Linux CPU names seem to be quite different. I'm not sure how to map them.

**Comment by Vexu** [comment]
If LLVM doesn't have the CPU you set `llvm_name` to null, for the `zig_name` looks like you just need to replace commas with underscores.

File: lib/std/zig/system/linux.zig [hunk 1]
```diff
@@ -344,6 +391,9 @@ pub fn detectNativeCpuAndFeatures() ?Target.Cpu {
         .powerpc, .powerpcle, .powerpc64, .powerpc64le => {
             return PowerpcCpuinfoParser.parse(current_arch, f.reader()) catch null;
         },
+        .riscv64, .riscv32 => {
```
Comment (kubkon) [inline lib/std/zig/system/linux.zig:394]: Is `riscv32` like `aarch64_32`?

Comment (RossComputerGuy) [reply lib/std/zig/system/linux.zig:394]: I'm not sure. I haven't heard of any 32-bit RISC-V systems which can run Linux as well.

Comment (kubkon) [reply lib/std/zig/system/linux.zig:394]: I'm asking to learn if we should collapse `riscv32` and `riscv64` into one prong but perhaps that is OK.

Comment (RossComputerGuy) [reply lib/std/zig/system/linux.zig:394]: Oh, I'm not sure heh.

**Comment by RossComputerGuy** [comment]
At this point with how close 0.12 is, I'd just say merge. This PR implements what it intends to. Another thing to note is that Linux on RISC-V is pretty much mainly common with the U74 and I haven't seen other systems with a different core. This means this PR will make CPU detection work on the vast majority of RISC-V systems running Linux.

**Pull request merged**